core/network/impl/gossiper_broadcast.cpp

```diff
@@ -5,87 +5,42 @@
 
 #include "network/impl/gossiper_broadcast.hpp"
 
+#include <atomic>
+#include <memory>
+
 #include "network/common.hpp"
-#include "network/helpers/scale_message_read_writer.hpp"
 #include "network/impl/loopback_stream.hpp"
 
 namespace kagome::network {
+  KAGOME_DEFINE_CACHE(stream_engine);
 
   GossiperBroadcast::GossiperBroadcast(libp2p::Host &host)
-      : host_{host}, logger_{common::createLogger("GossiperBroadcast")} {}
+      : logger_{common::createLogger("GossiperBroadcast")},
+        stream_engine_{StreamEngine::create(host)} {}
 
   void GossiperBroadcast::reserveStream(
       const libp2p::peer::PeerInfo &peer_info,
+      const libp2p::peer::Protocol &protocol,
       std::shared_ptr<libp2p::connection::Stream> stream) {
-    reserved_streams_.emplace(peer_info, std::move(stream));
+    stream_engine_->addReserved(peer_info, protocol, std::move(stream));
   }
 
-  void GossiperBroadcast::addStream(
-      std::shared_ptr<libp2p::connection::Stream> stream) {
-    auto peer_id_res = stream->remotePeerId();
-    if (not peer_id_res.has_value()) {
-      logger_->error("Can't get peer_id: {}", peer_id_res.error().message());
-      return stream->reset();
-    }
-    auto &peer_id = peer_id_res.value();
-
-    auto syncing_stream_it = syncing_streams_.find(peer_id);
-    if (syncing_stream_it != syncing_streams_.end()) {
-      if (syncing_stream_it->second != stream) {
-        syncing_stream_it->second->reset();
-        syncing_stream_it->second.swap(stream);
-        logger_->debug("Syncing stream (peer_id={}) was replaced",
-                       peer_id.toHex());
-      }
-      return;
-    }
-
-    logger_->debug("Syncing stream (peer_id={}) was emplaced", peer_id.toHex());
+  void GossiperBroadcast::storeSelfPeer(
+      const libp2p::peer::PeerInfo &peer_info) {
+    peer_info_ = peer_info;
```

Comment:
Er, AFAIK a peer is someone you are talking to, so this may be quite confusing. I suggest to rename it to storeSelfPeerInfo and the corresponding field to self_info_.